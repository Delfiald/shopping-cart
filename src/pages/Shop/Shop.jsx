import ShopMain from "../../components/Main/ShopMain";
import Aside from "../../components/Aside/Aside";
import { useOutletContext, useSearchParams } from "react-router-dom";
import { useEffect } from "react";

function Shop() {
 const [searchParams, setSearchParams] = useSearchParams();
 const category = searchParams.get("category");
 const { categories, products, hoverButton, setHoverButton } =
  useOutletContext();

 const page = searchParams.get("page");
 const itemPerPage = searchParams.get("itemsPerPage");
 const sort = searchParams.get("sort");
 const search = searchParams.get("search") || "";

 const displayedProducts = products
  .filter((product) => {
   if (category && category !== "all") {
    return product.category === category;
   }
   return true;
  })
  .filter((product) => {
   if (search) {
    return product.title
     .toLowerCase()
     .replace(/\s+/g, "")
     .includes(search.toLowerCase());
   }
   return true;
  });

 const handleFilterChange = (newCategory) => {
  setSearchParams((prevParams) => {
   const params = new URLSearchParams(prevParams);
   params.set("category", newCategory);
   params.set("page", 1);
   return params;
  });
 };

 const handlePageChange = (page) => {
  setSearchParams((prevParams) => {
   const params = new URLSearchParams(prevParams);
   params.set("page", page);
   return params;
  });
 };

 const handleItemsPerPageChange = (itemsPerPage) => {
  setSearchParams((prevParams) => {
   const params = new URLSearchParams(prevParams);
   params.set("page", 1);
   params.set("itemsPerPage", itemsPerPage);

   return params;
  });
 };

 const handleSortChange = (sort) => {
  setSearchParams((prevParams) => {
   const params = new URLSearchParams(prevParams);
   params.set("sort", sort);
   return params;
  });
 };

 useEffect(() => {
  const validCategory = categories.includes(category) ? category : "all";
  const validSort = [
   "name-asc",
   "name-desc",
   "price-asc",
   "price-desc",
  ].includes(sort)
   ? sort
   : "name-asc";
  const totalItems = displayedProducts.length;
  const maxPage = Math.ceil(totalItems / (itemPerPage || 10));
  const validPage = Math.max(1, Math.min(parseInt(page) || 1, maxPage));
  const validItemsPerPage = [5, 10, null].includes(parseInt(itemPerPage))
   ? itemPerPage
   : 10;

  const defaultParams = {
   category: validCategory,
   sort: validSort,
   page: validPage,
   itemsPerPage: validItemsPerPage,
   search: search,
  };

  const updatedParams = { ...defaultParams };

  if (
   JSON.stringify(updatedParams) !==
   JSON.stringify(Object.fromEntries(searchParams))
  ) {
   setSearchParams(updatedParams);
  }
 }, [
  categories,
  category,
  displayedProducts.length,
  itemPerPage,
  page,
  products,
  products.length,
  search,
  searchParams,
  setSearchParams,
  sort,
 ]);

 return (
  <>
   <Aside
    categories={categories}
    category={category}
    handleFilterChange={handleFilterChange}
   />
   <ShopMain
    category={category}
    products={displayedProducts}
    page={parseInt(page)}
    setPage={handlePageChange}
    itemPerPage={parseInt(itemPerPage)}
    setItemPerPage={handleItemsPerPageChange}
    sort={sort}
    setSort={handleSortChange}
    hoverButton={hoverButton}
    setHoverButton={setHoverButton}
    search={search}
   />
  </>
 );
}

export default Shop;
