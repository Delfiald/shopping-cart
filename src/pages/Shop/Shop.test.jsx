import { MemoryRouter, Outlet, Route, Routes } from "react-router-dom";
import Shop from "./Shop";
import { render } from "@testing-library/react";

const mockCategories = ["category-1", "category-2"];

const MockRouter = () => {
 return (
  <MemoryRouter initialEntries={["/shop"]}>
   <Routes>
    <Route
     path="/"
     element={<Outlet context={{ categories: mockCategories }} />}
    >
     <Route path="shop" element={<Shop />} />
    </Route>
   </Routes>
  </MemoryRouter>
 );
};

describe("Test Home Page", () => {
 it("Should Render Home Page", () => {
  const { container } = render(<MockRouter />);

  expect(container).toMatchSnapshot();
 });
});
