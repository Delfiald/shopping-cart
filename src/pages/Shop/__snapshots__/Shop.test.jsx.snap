// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test Home Page > Should Render Home Page 1`] = `
<div>
  <aside>
    <div
      class="_categories_1c2789"
    >
      <h2>
        Categories
      </h2>
      <div
        class="_category-lists_1c2789"
        data-testid="category-lists"
      >
        <button
          class="_category_1c2789"
          data-testid="category-button-0"
        >
          <div
            class="_displayed_1c2789"
            data-testid="category-1"
          >
            Category-1
          </div>
          <div
            class="_hovered_1c2789"
          >
            Category-1
          </div>
        </button>
        <button
          class="_category_1c2789"
          data-testid="category-button-1"
        >
          <div
            class="_displayed_1c2789"
            data-testid="category-2"
          >
            Category-2
          </div>
          <div
            class="_hovered_1c2789"
          >
            Category-2
          </div>
        </button>
      </div>
    </div>
  </aside>
  <main
    class="_shop_716e15"
  >
    <h2>
      All Products
    </h2>
    <div
      class="_product-list-header_716e15"
    >
      <div
        data-testid="list-header"
      >
        <div
          class="_item-information_716e15"
        >
          Showing 1 - 5 products
        </div>
        <div
          class="_option-wrapper_716e15"
        >
          <div
            class="_sort-button_716e15"
            data-testid="sort-button"
          >
            <div
              data-testid="sort-value"
            >
              <p>
                Name Asc
              </p>
              <svg
                class="lucide lucide-chevron-down"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m6 9 6 6 6-6"
                />
              </svg>
            </div>
          </div>
          <div
            class="_item-per-page_716e15"
            data-testid="item-per-page"
          >
            <div
              class="_item-per-page-value_716e15"
              data-testid="item-per-page-value"
            >
              <p>
                10
              </p>
              <svg
                class="lucide lucide-chevron-down"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m6 9 6 6 6-6"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <section
      class="_product-list-wrapper_716e15"
      data-testid="product-card"
    >
      <a
        class="_card_716e15"
        data-discover="true"
        data-testid="product-card-1"
        href="/product/1"
      >
        <div
          class="_image-wrapper_716e15"
        >
          <img
            alt="product-1"
            src="/image/product-1.png"
          />
        </div>
        <div
          class="_product-name_716e15"
        >
          product-1
        </div>
        <div
          class="_product-price_716e15"
        >
          $110
        </div>
      </a>
      <a
        class="_card_716e15"
        data-discover="true"
        data-testid="product-card-2"
        href="/product/2"
      >
        <div
          class="_image-wrapper_716e15"
        >
          <img
            alt="product-2"
            src="/image/product-2.png"
          />
        </div>
        <div
          class="_product-name_716e15"
        >
          product-2
        </div>
        <div
          class="_product-price_716e15"
        >
          $120
        </div>
      </a>
      <a
        class="_card_716e15"
        data-discover="true"
        data-testid="product-card-3"
        href="/product/3"
      >
        <div
          class="_image-wrapper_716e15"
        >
          <img
            alt="product-3"
            src="/image/product-3.png"
          />
        </div>
        <div
          class="_product-name_716e15"
        >
          product-3
        </div>
        <div
          class="_product-price_716e15"
        >
          $130
        </div>
      </a>
      <a
        class="_card_716e15"
        data-discover="true"
        data-testid="product-card-4"
        href="/product/4"
      >
        <div
          class="_image-wrapper_716e15"
        >
          <img
            alt="product-4"
            src="/image/product-4.png"
          />
        </div>
        <div
          class="_product-name_716e15"
        >
          product-4
        </div>
        <div
          class="_product-price_716e15"
        >
          $140
        </div>
      </a>
      <a
        class="_card_716e15"
        data-discover="true"
        data-testid="product-card-5"
        href="/product/5"
      >
        <div
          class="_image-wrapper_716e15"
        >
          <img
            alt="product-5"
            src="/image/product-5.png"
          />
        </div>
        <div
          class="_product-name_716e15"
        >
          product-5
        </div>
        <div
          class="_product-price_716e15"
        >
          $150
        </div>
      </a>
    </section>
    <div
      class="_product-list-bottom_716e15"
    >
      <div
        class="_pagination_716e15"
      >
        <div
          class="_arrow-button_716e15"
        />
        <div
          class="_number-button_716e15"
        >
          <div
            class="_current-page-button_716e15 _active_716e15"
            data-testid="current-page-button"
          >
            1
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
